### PR TITLE
feat: Improved instrumentation around core functions

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

We think there are some bottlenecks in prod but its unclear where they are. A few simple instrumentations would help to understand this

## Changes

* Downgrades some logs that we don't care about
* Instrument the main ingestion loop.
* Only stringifying the event if the timeout fires.
* Lock on hostname instead of uuid so that we can identify which pod is the cause

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: plugin-server/src/main/utils.ts [hunk 1]
```diff
@@ -4,29 +4,29 @@ import { exponentialBuckets, Histogram } from 'prom-client'
...
+    timeoutContext,
     func,
     statsKey,
     teamId,
-}: FunctionInstrumentation<T, E>): Promise<T> {
-    const timeout = timeoutGuard(timeoutMessage, {
-        event: JSON.stringify(event),
-    })
+}: FunctionInstrumentation<T>): Promise<T> {
+    const t = timeoutGuard(timeoutMessage ?? `Timeout warning for '${statsKey}'!`, timeoutContext, timeout)
```
Comment (tiina303) [inline plugin-server/src/main/utils.ts:32]: nit: why the change to single letter?

File: plugin-server/src/main/utils.ts [hunk 2]
```diff
@@ -4,29 +4,29 @@ import { exponentialBuckets, Histogram } from 'prom-client'
...
     timeoutMessage,
-    event,
+    timeout,
+    timeoutContext,
     func,
     statsKey,
     teamId,
-}: FunctionInstrumentation<T, E>): Promise<T> {
-    const timeout = timeoutGuard(timeoutMessage, {
-        event: JSON.stringify(event),
```
Comment (pauldambra) [inline plugin-server/src/main/utils.ts:23]: nice catch 🙌

File: plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer-v2.ts [hunk 3]
```diff
@@ -321,68 +321,86 @@ export class SessionRecordingIngesterV2 {
                     await this.partitionLocker.claim(messages)
                 }
 
-                for (const message of messages) {
-                    const { partition, offset, timestamp } = message
+                await runInstrumentedFunction({
+                    statsKey: `recordingingester.handleEachBatch.parseKafkaMessages`,
+                    func: async () => {
+                        for (const message of messages) {
```
Comment (tiina303) [inline plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer-v2.ts:327]: nit: I'd find this more readable if this was a function

File: plugin-server/src/utils/db/utils.ts [hunk 4]
```diff
@@ -30,12 +30,13 @@ export function sanitizeEventName(eventName: any): string {
 
 export function timeoutGuard(
     message: string,
-    context?: Record<string, any>,
+    context?: Record<string, any> | (() => Record<string, any>),
```
Comment (tiina303) [inline plugin-server/src/utils/db/utils.ts:33]: why this change? I didn't see it being used anywhere?

**Review by pauldambra — APPROVED** [review]
🚀

File: plugin-server/src/main/utils.ts [hunk 5]
```diff
@@ -20,18 +27,27 @@ export async function runInstrumentedFunction<T>({
     func,
     statsKey,
     teamId,
+    logToConsole = false,
```
Comment (tiina303) [inline plugin-server/src/main/utils.ts:30]: this is a confusing env name as timeoutGuard also logs to console, you'd expect it not to do that. So maybe `logExecutionTimes`

File: plugin-server/src/utils/db/utils.ts (hunk 4 above)
Comment (benjackwhite) [reply plugin-server/src/utils/db/utils.ts:33]: Called by the run function. Means we can generate it _only_ when the log is actually executed. Currently we call JSON.stringify() on every event for no reason

File: plugin-server/src/main/utils.ts (hunk 5 above)
Comment (benjackwhite) [reply plugin-server/src/main/utils.ts:30]: Fair point 👍

**Review by tiina303 — APPROVED** [review]
